docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,398 @@
+
+```toml
+[Writable.InsecureSecrets]
+    [Writable.InsecureSecrets.DB]
+    	path = "redisdb"
+    	[Writable.InsecureSecrets.DB.Secrets]
+            username = ""
+            password = ""
+```
+
+##### Handling on-the-fly changes to `InsecureSecrets`
+
+All services will need to handle the special processing when `InsecureSecrets` are changed on-the-fly via Consul. Since this will now be a common configuration item in `Writable` it can be handled in `go-mod-bootstrap` along with existing log level processing. This special processing will be taken from App SDK.
+
+#### Mocks
+
+Proper mock of the `SecretProvider` interface will be created with `Mockery` to be used in unit tests. Current mock in App SDK is handle written rather then generated with `Mockery`.
+
+#### Where will `SecretProvider` reside?
+
+##### Go Services
+
+The final decision to make is where will this new `SecretProvider` abstraction reside? Originally is was assumed that it would reside in `go-mod-secrets`, which seems logical. If we were to attempt this with the implementation including the bootstrap handler, `go-mod-secrets` would have a dependency on `go-mod-bootstrap` which will likely create a circular dependency. 
+
+Refactoring the existing implementation in `go-mod-bootstrap` and have it reside there now seems to be the best choice.
+
+##### C Device Service
+
+The C Device SDK will implement the same `SecretProvider` abstraction, InsecureSercets configuration and the underling `SecretStore` client.
+
+### Consequences
+
+- All service's will have `Writable.InsecureSecrets` section added to their configuration
+- `InsecureSecrets` definition will be moved from App SDK to go-mod-bootstrap
+- C SDK implementation could be big lift?
+- edgex-go services will be modified to use the single `SecretProvider` interface from the DIC in place of current usage of the `GetDatabaseCredentials` and `GetCertificateKeyPair` interfaces.
+  - Calls to `GetDatabaseCredentials` and `GetCertificateKeyPair` will be replaced with calls to `GetSecrets` API and appropriate processing of the returned secrets will be added. 
+- App SDK will be modified to use `GetSecrets` API in place of the `GetDatabaseCredentials` API
+- App SDK will be modified to use the new `SecretProvider` bootstrap handler
+- app-service-configurable's configuration profiles will be updated to remove the `SecretStoreExclusive` configuration and just use the existing `SecretStore` configuration
+- security-secretstore-setup will be enhanced as described in the [Exclusive Secret Stores only](#exclusive-secret-stores-only) section above
+- Adding new services that need known secrets added to their `SecretStore` requires stopping and restarting all the services. The is because security-secretstore-setup will no longer know the values for the know secrets after it has completed setup.
```

Comment:
Isn't the issue the security-secretstore-setup is no longer running (or waiting on /dev/null in docker) in order to handle the copy?

Reply:
Correct. It has finished and needs to rerun which then hoses the tokens and credentials for the other services.